model_analyzer/config/generate/neighborhood.py

```diff
@@ -228,38 +227,129 @@ def _get_visited_coordinates(self) -> List[Coordinate]:
+                home_measurement=home_measurement,
+                vectors=vectors,
+                measurements=measurements)
+        return step_vector
+
+    def _optimize_for_better_objectives(self,
+                                        home_measurement: RunConfigMeasurement,
+                                        vectors: List[Coordinate],
+                                        measurements: List[RunConfigMeasurement]
+                                        ) -> Coordinate:
+        """
+        Calculate a step vector that maximizes the current objectives.
+        If no vectors are provided, return zero vector.
+
+        Parameters
+        ----------
+        vectors
+            list of vectors from home coordinate to the neighboring
+            coordinates that are passing constraints
+        measurements
+            list of measurements of the neighboring coordinates that
+            are passing constraints
+
+        Returns
+        -------
+        step_vector
+            a coordinate that tells the direction to move.
+        """
+        step_vector = Coordinate([0] * self._config.get_num_dimensions())
+
+        if not vectors:
+            return step_vector
+
         for vector, measurement in zip(vectors, measurements):
-            if measurement:
-                weight = home_measurement.compare_measurements(measurement)
-            else:
-                # Minimize the effect of None measurement on the step vector.
-                weight = 0.0
+            weight = home_measurement.compare_measurements(measurement)
+            step_vector += vector * weight
+
+        step_vector /= len(vectors)
+        return step_vector
+
+    def _optimize_for_passing_constraints(self,
+                                          home_measurement: RunConfigMeasurement,
+                                          vectors: List[Coordinate],
+                                          measurements: List[RunConfigMeasurement]
+                                          ) -> Coordinate:
+        """
+        Calculate a step vector that steps toward the coordinates that
+        pass the constraints (set default weights to 1.0). When no vectors
+        are provided (meaning there are no neighbors passing constraints),
+        continue with the neighbors that are not passing constraints by
+        comparing how much they are close to passing constraints.
+
+        Parameters
+        ----------
+        vectors
+            list of vectors from home coordinate to the neighboring
+            coordinates that are passing constraints
+        measurements
+            list of measurements of the neighboring coordinates that
+            are passing constraints
+
+        Returns
+        -------
+        step_vector
+            a coordinate that tells the direction to move.
+        """
+        step_vector = Coordinate([0] * self._config.get_num_dimensions())
+
+        if not vectors:
```

Comment:
Either all of the code to determine the vectors should be inside the function, or all should be outside. Currently half is outside and half is inside.

Reply:
I agree. Vectors and measurements are now handled inside the function.